Add BibleNLP dataset

davidstap opened the pull request:
<!-- If you are not submitting for a dataset, feel free to remove the content below  -->


<!-- add additional description, question etc. related to the new dataset -->

## Checklist for adding MMTEB dataset

<!-- 
Before you commit here is a checklist you should complete before submitting
if you are not 
 -->
This dataset has an extremely large coverage of 829 languages, and includes numerous low-resource languages. For this reason I only considered English-centric directions, resulting in a reasonable 1656 pairs. I will add performance results later, but initial testing suggests that while quality is not good for most (low-resource) languages, it is far better than random guesses, so IMO a good addition to the benchmark.

- [x] I have tested that the dataset runs with the `mteb` package.
- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb run -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [x] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [x] I have added points for my submission to the [points folder](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) using the PR number as the filename (e.g. `438.jsonl`).

davidstap commented:
Could you take a look at this PR  @KennethEnevoldsen @imenelydiaker? I'm personally quite excited by the large language coverage.

isaac-chung commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+    "zyp_Latn",  # Zyphe Chin
+]
+
+_SPLIT = ["train"]
```
Probably good to note why the train split is chosen here since there are also validation and test splits a well.

davidstap replied:
Makes sense. I went for the train split because the validation/test splits are extremely small in a lot of cases (think 20 sentences). Given this I thought it would be better to subsample the train set.

isaac-chung commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+        dataset={
+            "path": "davidstap/biblenlp-corpus-mmteb",
+            "revision": "264a18480c529d9e922483839b4b9758e690b762",
+            "split": f"train[:{_N}]",
```
Cool - I wasn't familiar with this convention.

isaac-chung reviewed (commented):
This is awesome! Just a few comments.

dokato commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+        eval_splits=_SPLIT,
+        eval_langs=_LANGUAGES_MAPPING,
+        main_score="f1",
+        date=("2023-04-09", "2023-04-09"),
```
Here is an interesting one, cause [for consistency](https://github.com/embeddings-benchmark/mteb/pull/545#discussion_r1577555935) it should be the date of when the source was written 🤔 or maybe in this case a period when the translations where made if such information available? One way or another it requires some adnotation in the comment.

davidstap replied:
I see, interesting indeed. For English, I mostly used the World English Bible (WEB) versions. On Wikipedia I can find: first draft in 1997, completed in 2020, so I will add this information. 

It seems impossible, or at the minimum infeasible, to add information for all other languages.

dokato commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+def extend_lang_pairs_english_centric() -> dict[str, list[str]]:
+    # add all language pairs with english as source or target
+    hf_lang_subset2isolang = {}
+    for x in _LANGUAGES:
```
if we know that one of the languages is english, we don't need to iterate over twice.
consider sth like:
```py
for lang in _LANGUAGES:
  pair = f"eng_Latn-{lang}"
  hf_lang_subset2isolang[pair] = ['eng_Latn', lang.replace("_", "-")]
  pair = f"{lang}-eng_Latn"
  hf_lang_subset2isolang[pair] = [lang.replace("_", "-"), 'eng_Latn']

```

davidstap replied:
You're right, thanks for catching. I lazily copied this over from some other task, will change to your suggestion.

dokato commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+        n_samples={"train": _N},
+        avg_character_length={"train": 120},
+        bibtex_citation="""
+@article{akerman2023ebible,
```
if you improve indentation here

dokato commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+
+    def dataset_transform(self) -> None:
+        # Convert to standard format
+        for lang in self.langs:
```
if we edit load data already, I don't see why we would need to iterate over `self.langs` twice? maybe we don't need `dataset_transform` at all?

davidstap replied:
Sure we can do the transformation in `load_data` to prevent iterating twice, I'll change it.

davidstap replied:
On second thought: This would not be an improvement because it requires different transformations for if else. (The if else is there to prevent double loading of e.g. 'aai_Latn-eng_Latn' and 'eng_Latn-aai_Latn', which is the same huggingface data, and loading data is slow #572).

IMO the current code is a better trade-off between readability and efficiency. The impact of the second loop is negligible; it is a very cheap operation over 829*2 items.

dokato commented on mteb/tasks/BitextMining/multilingual/BibleNLPBitextMining.py:
```diff
@@ -0,0 +1,955 @@
+    @staticmethod
+    def _transform_lang_name_hf(lang):
+        # Transform language name to match huggingface configuration
+        langs = [l.split("_")[0] for l in lang.split("-")]
```
if we want to detect english, this looks like a regex job to me

dokato reviewed (commented):
Really interesting dataset :) find some of my suggestions below

davidstap commented:
I have addressed your comments @isaac-chung @dokato, thanks for your suggestions!

Results are included now as well. `intfloat/multilingual-e5-small` performs best, but like I pointed out the performance is bad for most low-resource languages. However, almost all directions are way better than random chance (1/256), so this seems like a valuable addition to evaluate (future) multilingual models.

I think this is ready to merge now, do you agree? Let me know, then I'll calculate the points. Thanks!

isaac-chung reviewed (approved):
Thanks for iterating! I think it's good enough to merge. In terms of points, the dataset is 2 + 4 * number of new languages added to the BitextMining task. Please check if any of the languages have already been covered in that folder and the multilingual subfolder as well. Then 2 pts each to @dokato and I for the reviews.

davidstap commented:
Thanks for your quick response! I calculated the number of new languages using [this script](https://gist.github.com/davidstap/787acb00fecb24559439b08cac30c5ea): 756 new languages.

I'll update the points and merge.

davidstap commented:
Ready to merge now @isaac-chung (I don't have write access)
